cnd/src/swap_protocols/halight/connector.rs

```diff
@@ -33,39 +38,77 @@ enum InvoiceState {
 // ref: https://api.lightning.community/#paymentstatus
 #[derive(Copy, Clone, Debug, Deserialize, PartialEq)]
 #[serde(rename_all = "SCREAMING_SNAKE_CASE")]
-enum PaymentStatus {
+pub enum PaymentStatus {
     Unknown,
     InFlight,
     Succeeded,
     Failed,
 }
 
-#[derive(Debug, Deserialize)]
-struct Invoice {
-    pub value: String,
-    pub value_msat: String,
-    pub amt_paid_sat: String,
-    pub amt_paid_msat: String,
-    pub expiry: String,
-    pub cltv_expiry: String,
+trait ValidateParams {
+    fn validate(self, params: halight::Params) -> Result<(), ValidationError>;
+}
+
+#[derive(Clone, Copy, Debug, thiserror::Error)]
+pub enum ValidationError {
+    #[error("Params do not match lnd invoice: (expected {0:?}, got {1:?})")]
+    Invoice(halight::Params, Invoice),
+    #[error("Params do not match lnd payment: (expected {0:?}, got {1:?})")]
+    Payment(halight::Params, Payment),
+}
+
+#[derive(Clone, Copy, Debug, Deserialize)]
+pub struct Invoice {
+    #[serde(deserialize_with = "deserialize_amount")]
+    pub value: asset::Bitcoin,
+    #[serde(deserialize_with = "deserialize_timestamp")]
+    pub expiry: Timestamp,
+    #[serde(deserialize_with = "deserialize_timestamp")]
+    pub cltv_expiry: Timestamp,
     pub state: InvoiceState,
+    pub fallback_addr: ethereum::Address,
     #[serde(deserialize_with = "deserialize_r_preimage")]
     pub r_preimage: Option<[u8; 32]>,
 }
 
+impl ValidateParams for Invoice {
+    fn validate(self, params: halight::Params) -> Result<(), ValidationError> {
+        if params.lightning_cltv_expiry == self.cltv_expiry
```

Comment:
I am still not sure about these checks particularly the fallback_addr and lightning_amount